website/blog/2025-07-23-dynamic-database-credentials-with-hashicorp-vault-and-ferretdb.md

```diff
@@ -0,0 +1,209 @@
+- **Centralized management:** Manage database credentials from a single, secure Vault instance.
+
+## Connecting HashiCorp Vault to FerretDB
+
+Connecting HashiCorp Vault's database secrets engine to your FerretDB instance involves configuring Vault to use the MongoDB plugin and providing the FerretDB connection details.
+
+Here's a step-by-step guide to get you started with a local Vault and FerretDB setup:
+
+1. **Ensure FerretDB is running:** Make sure your FerretDB instance is active and accessible, with authentication.
+   If you haven't set it up yet, refer to our [FerretDB Installation Guide](https://docs.ferretdb.io/installation/ferretdb/).
+   You need a user with sufficient privileges to create, update, and revoke other users.
+   This will be the "root user" Vault uses for its operations.
+
+2. **Set up and initialize HashiCorp Vault:**
+   For local testing, you can run Vault in dev mode.
+   This will start a single-node Vault server that is unsealed and initialized.
+   You can run Vault in a Docker container:
+
+   ```sh
+   docker run -d --name vault -p 8200:8200 \
+   --cap-add=IPC_LOCK \
+   -e 'VAULT_DEV_ROOT_TOKEN_ID=<root-token>' \
+   -e 'VAULT_DEV_LISTEN_ADDRESS=0.0.0.0:8200' \
+   hashicorp/vault:latest
+   ```
+
+   `<root-token>` should have the value for your development root token ID – replace it with any string you prefer for easy access during development.
+
+   You need the Vault CLI to interact with the Vault server.
+   Ensure that you have the Vault CLI installed and configured to PATH on your system.
+   If you haven't installed it yet, you can follow the [Vault installation guide](https://developer.hashicorp.com/vault/tutorials/get-started/install-binary) to set it up.
+
+   Set the `VAULT_ADDR` environment variable:
+
+   ```sh
+   export VAULT_ADDR='http://127.0.0.1:8200'
+   ```
+
+   Log in to Vault using the root token you set during the `docker run` command:
+
+   ```sh
+   vault login <root-token>
+   ```
+
+   You should see a success message confirming your login.
+
+3. **Enable the database secrets engine:**
+   Enable the database secrets engine at a desired path, for example, `ferretdb-creds`.
+   This path will serve as the base for all database-related operations in Vault.
+
+   ```sh
+   vault secrets enable -path=ferretdb-creds database
+   ```
+
+4. **Configure the MongoDB plugin for FerretDB:**
+   Configure the `ferretdb-creds` secrets engine to use the mongodb-database-plugin and provide the connection details for your FerretDB instance.
+   The `connection_url` will use a templated format for enhanced security and to enable root credential rotation.
+
+   ```sh
+   vault write ferretdb-creds/config/ferretdb-conn \
+   plugin_name=mongodb-database-plugin \
+   allowed_roles="my-app-role" \
+   connection_url="mongodb://{{username}}:{{password}}@host.docker.internal:27017/" \
+   username="<ferretdb_username>" \
+   password="<ferretdb_password>"
```

Comment:
> You need a user with sufficient privileges to create, update, and revoke other users. This will be the "root user" Vault uses for its operations.

Could you add a comment to replace `<ferretdb_username>` with what you previously referred as "root user" please